### ISSUE TITLE
samtools command should use the bam index to speed up operations

### ISSUE THREAD

**Issue by jeanlain** [samtools/samtools#1038]
#### Is your feature request related to a problem? Please specify.
The samtools commands that take a bed file as argument to work on selected genomic regions or positions , such as `view`, `depth` and `mpileup`, do not take advantage of bam indexes (only to retrieve a single region), resulting in very slow operations on large files.
For instance, `samtools view` on a 40GB bam file and ~5000 regions from various contigs takes 12 min to run, while the equivalent command launched from Rsamtools (which uses the bam index) takes 13 s.

#### Describe the solution you would like.
These command should either detect the bam index automatically or take it as argument, and use it to speed up operations.
This is related to https://github.com/samtools/samtools/issues/329.
It was commented that my suggestion had not been implemented due to RAM usage concerns, which seems anachronic as the size of a bam index is very small compared to typical system RAM. The issue was flagged "urgent" in 2016.

Thanks.

**Comment by valeriuo** [comment]
Yes, the BED file is used as a post-processing filter, while the index is required when regions are explicitly specified as command line arguments. 
For `samtools view`, there is however a way to couple the BED file with an index by using the multi-region iterator option (`-M`). Just out of curiosity, what is the running time when using the `-M` option with `view`? 

`samtools view -ML <bed_file> <bam_file>`

**Comment by jeanlain** [comment]
Ok did didn't know about option `-M`. :)
With the command you suggest, samtools is now quite a bit faster than Rsamtools. So I guess that, at least for `view`, my suggestion is already implemented.

**Comment by diekhans** [comment]
Using the -M option with -L is not obvious from the documentation.  Mentioning it will the -L help message would address this issue.

**Comment by jkbonfield** [comment]
Agreed.  I think it needs clarifying across the board.   Even the original and traditional sense of -L for streaming-filter vs regions for random access isn't documented anywhere, and -M adds another layer to that.  We just haven't got around to doing it yet.

See also https://github.com/samtools/samtools/issues/1232
